banyan/src/forest/mod.rs

```diff
@@ -41,17 +41,30 @@ impl<T: TreeTypes> BranchCache<T> {
 /// There might be more types in the future, so this essentially acts as a module for the entire
 /// code base.
 pub trait TreeTypes: Debug + Send + Sync + Clone + 'static {
-    /// key type. This also doubles as the type for a combination (union) of keys
+    /// key type
     type Key: Debug + Eq + Send;
+    /// Type for a summary of keys. In some cases this can be the same type as the key type.
+    type Summary: Debug + Eq + Send;
     /// compact sequence type to be used for indices
-    type Seq: CompactSeq<Item = Self::Key>
+    type KeySeq: CompactSeq<Item = Self::Key>
         + Serialize
         + DeserializeOwned
         + Clone
         + Debug
         + FromIterator<Self::Key>
         + Send
-        + Sync;
+        + Sync
+        + Summarizable<Self::Summary>;
+    /// compact sequence type to be used for indices
+    type SummarySeq: CompactSeq<Item = Self::Summary>
+        + Serialize
+        + DeserializeOwned
+        + Clone
+        + Debug
+        + FromIterator<Self::Summary>
+        + Send
+        + Sync
```

Comment:
can the usual ones be factored out into an alias?

Reply:
I was thinking about that. What you can do is this:

```rust
trait TheUsualSuspects {}

impl <T: Debug + Send + Sync + ...> TheUsualSuspects {}
```

But then, it adds another level of indirection to the reader. Maybe the verbose but simple thing is better...